aten/src/ATen/native/Convolution.cpp

```diff
@@ -1810,7 +1810,10 @@ std::tuple<Tensor, Tensor, Tensor> convolution_backward(
     }
   }
 
-  return std::make_tuple(backend_grad_input, backend_grad_weight, backend_grad_bias);
+  return std::make_tuple(
+    output_mask[0] ? backend_grad_input : Tensor(),
+    output_mask[1] ? backend_grad_weight : Tensor(),
+    output_mask[2] ? backend_grad_bias : Tensor());
```

Comment:
I changed my mind on this. I think we should be putting this logic into https://github.com/pytorch/pytorch/blob/a8e45b596910f90013637ccc735d6ca21a93d852/aten/src/ATen/native/mkldnn/Conv.cpp#L207-L210 to fix the problem closer to the root. Bonus points if we can figure out what exactly is going wrong with the mkldnn path (but in any case, the fix is better than the current situation).

The reason why I changed my mind is: it should be valid for convolution_backward to return a correctly-shaped tensor full of zeros OR an undefined Tensor (undefined Tensor is autograd's optimization for "correctly-shaped Tensor full of zeros"). We (PyTorch) control most of the code inside the convolution_backward function, but there is a "convolution_backward_overridable" path where we let external parties add their own convolution that we don't control.

Reply:
I can see doing it at this outer level "hiding" the issue and there may end up being a lot of backends that return garbage for the masked out values. But maybe that is not an issue unless you consider each convolution backend e.g. mkldnn_convolution_backend to exist as stand-alone functions in their own right and must also abide by the same rules that the general convolution does.

Reply:
Right now I'm mostly worried about BC concerns for convolution_overridable_backward: we definitely did not advertise this spec (that convolution_backward must return undefined Tensor if the gradient isn't being computed) when we created convolution_overridable_backward. I don't know what the users of convolution_overridable_backward out there are doing but I don't want to break them if they are relying on returning a Tensor full of zeros rather than undefined tensor.

Reply:
Oh I see, I'm just not sure if it is possible for the user to observe that behavior (in backward). See `Functions.cpp` it doesn't look like `std::get<2>(grad_result)` will be copied into `grad_inputs` which is returned unless `output_mask[2]` is 1.
```
variable_list ConvolutionBackward0::apply(variable_list&& grads) {
  std::lock_guard<std::mutex> lock(mutex_);

  IndexRangeGenerator gen;
  auto input_ix = gen.range(1);
  auto weight_ix = gen.range(1);
  auto bias_ix = gen.range(1);
  variable_list grad_inputs(gen.size());
  const auto& grad = grads[0];
  auto input = input_.unpack();
  auto weight = weight_.unpack();
  if (should_compute_output({ input_ix, weight_ix, bias_ix })) {
      auto grad_input_mask = std::array<bool, 3>{
        should_compute_output({ input_ix }),
        should_compute_output({ weight_ix }),
        should_compute_output({ bias_ix }),
      };
    auto grad_result = grad.defined() ? convolution_backward(grad, input, weight, bias_sizes_opt, stride, padding, dilation, transposed, output_padding, groups, grad_input_mask) : std::tuple<Tensor, Tensor, Tensor>();
      if (should_compute_output({ input_ix })) {
        copy_range(grad_inputs, input_ix, std::get<0>(grad_result));
      }
      if (should_compute_output({ weight_ix })) {
        copy_range(grad_inputs, weight_ix, std::get<1>(grad_result));
      }
      if (should_compute_output({ bias_ix })) {
        copy_range(grad_inputs, bias_ix, std::get<2>(grad_result));
      }
  }
  return grad_inputs;
}
```

Reply:
Hmm maybe in double backward though...

Reply:
Not sure if it can even happen in double backward either, so the only scenario where there are bc concerns is when convolution_backward is called out of the context of the engine. Maybe that is an issue? cc @albanD